### PR TITLE
Desktop: preview mode fix

### PR DESCRIPTION
**Pull request by Runo-saduwa**
**DESCRIPTION:**
This patch is a fix for the issue #2467 . Previously before this solution, the Edit menu items were not disabled on PREVIEW, this caused users to be able to use short cuts such as CTRL+B, CTRL+` on PREVIEW mode which is an unwanted behavior.

**HOW TO TEST**
Use the following test cases to confirm this solution::

- Add a note to a notebook and click the layout toggler to PREVIEW:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- Add a note to a notebook and click the layout toggler to EDITOR:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Add a note to a notebook and click the layout toggler to EDITOR AND PREVIEW MODE:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Imported a HTML note  and click the layout toggler to EDITOR/PREVIEW MODE/SPLIT:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- CREATE an Empty Notebook (book without notes)
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**

### CODE REVIEW

File: ElectronClient/app.js [hunk 1]
```diff
@@ -1178,7 +1182,16 @@ class Application extends BaseApplication {
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			if (viewer) {
+				if (viewer === 'viewer') {
+					menuItem.enabled = false;
+				} else {
+					menuItem.enabled = true;
+				}
```
Comment (mic704b) [inline ElectronClient/app.js]: This can be one line  `menuItem.enabled = (viewer === 'viewer');

File: ElectronClient/app.js [hunk 2]
```diff
@@ -1178,7 +1182,16 @@ class Application extends BaseApplication {
...
+			} else {
+
```
Comment (mic704b) [inline ElectronClient/app.js]: unnecessary white space

File: ElectronClient/app.js [hunk 3]
```diff
@@ -281,6 +281,9 @@ class Application extends BaseApplication {
 
 		if (['NOTE_VISIBLE_PANES_TOGGLE', 'NOTE_VISIBLE_PANES_SET'].indexOf(action.type) >= 0) {
 			Setting.setValue('noteVisiblePanes', newState.noteVisiblePanes);
+			const viewer = newState.noteVisiblePanes[0];
+			this.updateMenuItemStates(viewer);
+
```
Comment (mic704b) [inline ElectronClient/app.js]: unnecessary white space

**Comment by mic704b** [comment]
@Runo-saduwa nice work, didn't check the logic, just a few style comments.

**Comment by Runo-saduwa** [comment]
I am still trying to set up Web Clipper to test this. as soon as i successfully set up Web Clipper on my browser, this issue will be resolved

**Comment by tessus** [comment]
Just use the "normal" web clipper with your production Joplin insgtance. Then create a HTML note as I explained earlier, export it (jex), and import it into your dev Joplin instance.

**Comment by tessus** [comment]
Or use this one:

[html-test.zip](https://github.com/laurent22/joplin/files/4253313/html-test.zip)

**Comment by Runo-saduwa** [comment]
Thank You, I will give updates shortly

**Comment by Runo-saduwa** [comment]
> Just use the "normal" web clipper with your production Joplin insgtance. Then create a HTML note as I explained earlier, export it (jex), and import it into your dev Joplin instance.

Special thanks to @tessus

**Comment by Runo-saduwa** [comment]
@PackElend Please label me, thanks.

**Comment by Runo-saduwa** [comment]
Thanks for the quick review @tessus

File: ElectronClient/app.js [hunk 4]
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
...
+	async updateMenuItemStates(viewer, state = null) {
 		if (!this.lastMenuScreen_) return;
 		if (!this.store() && !state) return;
 
 		if (!state) state = this.store().getState();
 
 		const selectedNoteIds = state.selectedNoteIds;
 		const note = selectedNoteIds.length === 1 ? await Note.load(selectedNoteIds[0]) : null;
-
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
```
Comment (laurent22) [inline ElectronClient/app.js]: Please remove empty line from commit.

File: ElectronClient/app.js [hunk 5]
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
...
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const noteCheck = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			if (noteCheck === false || viewer === 'viewer') {
+				menuItem.enabled = false;
+			} else if (noteCheck === true || viewer !== 'viewer') {
+				menuItem.enabled = true;
+			}
 		}
-
 		const menuItem = Menu.getApplicationMenu().getMenuItemById('help:toggleDevTools');
```
Comment (laurent22) [inline ElectronClient/app.js]: Please remove empty line from commit.

File: ElectronClient/app.js [hunk 6]
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
...
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
-			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const noteCheck = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			if (noteCheck === false || viewer === 'viewer') {
+				menuItem.enabled = false;
+			} else if (noteCheck === true || viewer !== 'viewer') {
+				menuItem.enabled = true;
+			}
```
Comment (laurent22) [inline ElectronClient/app.js]: Sorry I can't wrap my head around this logic. There's an implied "else" case - in which case is it executed? Never? Then why use an else/if?

Also please could you clarify what "noteCheck" means? Is it possible to find a name that better expresses your intent?

Comment (Runo-saduwa) [reply ElectronClient/app.js]: about the else/if statement, i have changed that to else and the logic still works fcine. noting was broken.

and the `noteCheck` variable is used to store the boolean value that will be returned from this statement `!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;`, which is used to check if there is a note and also if the note is a HTML note.

i have added a comment to explain that part.

Comment (mic704b) [reply ElectronClient/app.js]: Hi @Runo-saduwa, I don't want to interfere, but just to suggest maybe something like 'isHtmlNote' might be a better name for 'noteCheck'.  Or ideally maybe don't bother with making a temporary variable at all because you don't need it again.

Another style tip, consider using just 'if (x)' instead of if (x === true)' unless you have an explicit reason for doing the latter. Writing it this way will also help you think of a good name for it.

Comment (Runo-saduwa) [reply ElectronClient/app.js]: @mic704b that’s a great suggestion, thanks. I’ll with your suggestions. 

isHtmlNote is cool

Comment (Runo-saduwa) [reply ElectronClient/app.js]: The code has been improved again, this time I improved the variable name to something meaningful (isHtmlNote) and I also replaced  the If statement with a sweet ternary operator. 

You can go ahead and test it, noting is broken, everything is in order.

Comment (mic704b) [reply ElectronClient/app.js]: >  I also replaced the If statement with a sweet ternary operator

Better, but switch it around and it would be even better.

Comment (mic704b) [reply ElectronClient/app.js]: And remove the comment about `noteCheck`

Comment (Runo-saduwa) [reply ElectronClient/app.js]: Yeah Thanks, I just did that

File: ElectronClient/app.js [hunk 7]
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			!isHtmlNote || viewer === 'viewer' ?  menuItem.enabled = false : menuItem.enabled = true;
```
Comment (laurent22) [inline ElectronClient/app.js]: This is like writing `true ? x = false : x = true;` or you know, you could just write `x = false`. This is the same here, you could simply write `menuItem.enabled = !isHtmlNote && viewer !== 'viewer'`.

Comment (Runo-saduwa) [reply ElectronClient/app.js]: I just tried it, it doesn’t work as expected, it breaks the expected functionality. I think this is the simplest and most readable way to solve the issue and maintain its expectated behavior.

Comment (mic704b) [reply ElectronClient/app.js]: `menuItem.enabled = isHtmlNote && viewer !== 'viewer';`

Comment (laurent22) [reply ElectronClient/app.js]: ah, now I see what the problem is. We don't want the items to be enabled if it is an HTML note or if it the viewer. The problem here is that `isHtmlNote` is false when it's an HTML note (and true when it's a Markdown note), so Runo you have it backward.

Essentially you wrote this: `isHtmlNote = note.type === 'markdown'` - doesn't that tell you that it's in fact a Markdown note, and not an HTML one? So full condition should be `const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;`

Comment (laurent22) [reply ElectronClient/app.js]: And I insist on getting rid of `true ? x = false : x = true;`

Comment (mic704b) [reply ElectronClient/app.js]: Perfect!

> it doesn’t work as expected, it breaks the expected functionality

@Runa-saduwa whenever you think this, it means it's time to look closer to find out why!

Comment (Runo-saduwa) [reply ElectronClient/app.js]: > And I insist on getting rid of true ? x = false : x = true;

i apologize, I have figured it out. I’ve gotten rid this

Comment (Runo-saduwa) [reply ElectronClient/app.js]: @mic704b you’re very correct. Thank you for that, I was a bit hasty in trying to solve the issue that brought about the PR

Comment (mic704b) [reply ElectronClient/app.js]: All good, now make sure to read Laurent's comment above closely, and adjust the code as described if it makes sense.

Laurent said:
> Essentially you wrote this: `isHtmlNote = note.type === 'markdown'` - doesn't that tell you that it's in fact a Markdown note, and not an HTML one? So full condition should be `const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;`

Comment (Runo-saduwa) [reply ElectronClient/app.js]: I’ve made the adjustments.

If this isn’t okay, I still have another way to do this, I’ve looked closely at the code and I’ve seen different ways to achieve what we want.

File: ElectronClient/app.js [hunk 8]
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language !== MarkupToHtml.MARKUP_LANGUAGE_HTML;
+			menuItem.enabled = isHtmlNote && viewer !== 'viewer';
```
Comment (laurent22) [inline ElectronClient/app.js]: I was expecting I could merge this, but after all this time it's still not good. I don't know what to say since we've already went over this. If something's not clear about what should happen please ask.

Also please rename the "viewer" variable, it doesn't make sense. It's like having a variable called "plane" and setting it to "car".

Comment (Runo-saduwa) [reply ElectronClient/app.js]: My bad, I thought I understood your last comment well enough, that was why I didn’t want to bother you.

But based on what you said, I could change the `isHtmlNote` to `isMarkDownNote` And turn the final statement to:

```isMarkDownNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN```

With the above code,  when the note is a HTML note, `isMarkDownNote` will be `false`And when the note is a Markdown note, `isMarkDownNote` will be `true`.

How do you see this approach, is it okay?

Comment (Runo-saduwa) [reply ElectronClient/app.js]: I’ll change that to `editorMode`

Comment (tessus) [reply ElectronClient/app.js]: > I’ll change that to editorMode

This is also wrong. Dude, you have 3 layouts: viewer, split, editor. So how would you call the mode? I'd say `layout` or `layoutMode`.

Comment (tessus) [reply ElectronClient/app.js]: > How do you see this approach, is it okay?

No, it's not. You don't have to rename the variable `isHtmlNote`. Anyway, I'm too tired now. Ttyl.

**Comment by PackElend** [comment]
@Runo-saduwa can you summarize you changes?
Please also include the comments / reasons.
This PR has a lot of comments, so it is quite challenging to follow it.
thank you

**Comment by Runo-saduwa** [comment]
>  So full condition should be const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
based on this comment by Laurent, i have changed the code to use this.
``` 
const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML
```
i also renamed the `viewer` variable to `layout` because it connotes the meaning of the value it stores.

Then i modified the logic for enabling and disabling the `menuItem` with this condition 
```!isHtmlNote && layout !== 'viewer' && !!note```
which causes the same side-effects as this:
```
 if(isHtmlNote || layout === 'viewer' || !note){ 
        menuItem.enabled = false;
 } else {
	menuItem.enabled = true;
 }
```

After doing all these changes, i tested with the following cases and confirmed that it works properly:

- Add a note to a notebook and click the layout toggler to PREVIEW:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- Add a note to a notebook and click the layout toggler to EDITOR:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Add a note to a notebook and click the layout toggler to EDITOR AND PREVIEW MODE:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Imported a HTML note  and click the layout toggler to EDITOR/PREVIEW MODE/SPLIT:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- CREATE an Empty Notebook (book without notes)
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**

File: ElectronClient/app.js [hunk 9]
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
```
Comment (Runo-saduwa) [inline ElectronClient/app.js]: >  So full condition should be const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
based on this comment by Laurent, i have changed the code to use this.
``` 
const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML
```

File: ElectronClient/app.js [hunk 10]
```diff
@@ -1166,7 +1169,7 @@ class Application extends BaseApplication {
 		this.lastMenuScreen_ = screen;
 	}
 
-	async updateMenuItemStates(state = null) {
+	async updateMenuItemStates(layout, state = null) {
```
Comment (Runo-saduwa) [inline ElectronClient/app.js]: i also renamed the `viewer` variable to `layout` because it connotes the meaning of the value it stores.

File: ElectronClient/app.js [hunk 11]
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
+			menuItem.enabled = !isHtmlNote && layout !== 'viewer' && !!note;
```
Comment (Runo-saduwa) [inline ElectronClient/app.js]: Then i modified the logic for enabling and disabling the `menuItem` with this condition 
```!isHtmlNote && layout !== 'viewer' && !!note```
which causes the same side-effects as this:
```
 if(isHtmlNote || layout === 'viewer' || !note){ 
        menuItem.enabled = false;
 } else {
	menuItem.enabled = true;
 }
```

File: ElectronClient/app.js (hunk 8 above)
Comment (mic704b) [reply ElectronClient/app.js]: Hi @Runo-saduwa 

tessus is right about the layouts and I think you've sorted that out now.

Concerning your question about `isHtmlNote`, tessus is also right, the variable you have in the code above is correctly named.  

However, as you suggest, you could instead implement a `isMarkdownNote`.  Either variation will work. Which one you choose to go with depends on how you use it thereafter, which one makes the your following code simpler to read and understand.

Comment (tessus) [reply ElectronClient/app.js]: Hi @mic704b . Runo has already made all the requested changes. The code above is outdated. I think the best way is to just click on `Files changed` and start a new review.

It's a bit confusing that these conversations are not automarically resolved as soon as the code does not exist anymore.

Comment (mic704b) [reply ElectronClient/app.js]: Thanks @tessus, I'm a bit behind.  I just wanted to respond to Runo's question.  Although it wasn't resolving the issues that were being discussed on this thread, the use of `isMarkdownNote` was an OK option to consider for other reasons.

**Comment by tessus** [comment]
Anything left to do here or can we merge this?

**Comment by laurent22** [comment]
We got there eventually! Thanks @Runo-saduwa for the fix and everyone for reviewing.
